File: metrics/descriptors/src/main/kotlin/org/convos/metrics/descriptors/navigation/NavigationTargets.kt
package org.convos.metrics.descriptors.navigation

import org.convos.metrics.annotations.NavigationTarget

// Primary Navigators

@NavigationTarget
interface ConversationsNavigator {
    class Args

    fun navigateTo(conversation: ConversationNavigator.Args)
    fun present(appSettings: AppSettingsNavigator.Args)
    fun present(newConversation: NewConversationNavigator.Args)
    fun present(explodeConfirmation: ExplodeConfirmationNavigator.Args)
    fun present(connectionGrant: ConnectionGrantNavigator.Args)
    fun present(explodeInfo: ExplodeInfoNavigator.Args)
    fun present(pinLimitInfo: PinLimitInfoNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ConversationNavigator {
    data class Args(val conversationId: String)

    fun present(paywall: PaywallNavigator.Args)
    fun present(conversationInfo: ConversationInfoNavigator.Args)
    fun present(myInfo: MyInfoNavigator.Args)
    fun present(memberProfile: MemberProfileNavigator.Args)
    fun present(shareInvite: ShareInviteNavigator.Args)
    fun present(newConversation: NewConversationNavigator.Args)
    fun present(reactions: ReactionsNavigator.Args)
    fun present(explodeInfo: ExplodeInfoNavigator.Args)
    fun present(lockedConvoInfo: LockedConvoInfoNavigator.Args)
    fun present(fullConvoInfo: FullConvoInfoNavigator.Args)
    fun present(conversationForkedInfo: ConversationForkedInfoNavigator.Args)
    fun present(revealMediaInfo: RevealMediaInfoNavigator.Args)
    fun present(photosInfo: PhotosInfoNavigator.Args)
    fun present(assistantConfirmation: AssistantConfirmationNavigator.Args)
    fun present(assistantInfo: AssistantInfoNavigator.Args)
    fun present(processingPowerInfo: ProcessingPowerInfoNavigator.Args)
    fun present(explodedInviteInfo: ExplodedInviteInfoNavigator.Args)
    fun present(setupProfile: SetupProfileNavigator.Args)
    fun present(inviteAccepted: InviteAcceptedNavigator.Args)
    fun present(requestPushNotifications: RequestPushNotificationsNavigator.Args)
    fun present(backwardsSecrecyInfo: BackwardsSecrecyInfoNavigator.Args)

    fun closed(context: ScreenContext)
}

// Root-Level Modals

@NavigationTarget
interface AppSettingsNavigator {
    class Args

    fun navigateTo(myInfo: MyInfoNavigator.Args)
    fun navigateTo(customize: CustomizeSettingsNavigator.Args)
    fun navigateTo(assistants: AssistantSettingsNavigator.Args)
    fun navigateTo(connections: ConnectionsNavigator.Args)
    fun navigateTo(backupRestore: BackupRestoreNavigator.Args)
    fun navigateTo(deleteAllData: DeleteAllDataNavigator.Args)
    fun navigateTo(subscriptionSettings: SubscriptionSettingsNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface NewConversationNavigator {
    data class Args(
        val mode: NewConversationMode,
        val inviteCode: String? = null,
    )

    fun closed(context: ScreenContext)
}

enum class NewConversationMode {
    CREATE,
    SCANNER,
    JOIN_INVITE,
}

@NavigationTarget
interface ExplodeConfirmationNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

// Conversation Detail Modals

@NavigationTarget
interface ConversationInfoNavigator {
    data class Args(val conversationId: String)

    fun navigateTo(edit: ConversationInfoEditNavigator.Args)
    fun navigateTo(membersList: MembersListNavigator.Args)
    fun navigateTo(filesAndLinks: AssistantFilesLinksNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ConversationInfoEditNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface MembersListNavigator {
    data class Args(val conversationId: String)

    fun navigateTo(memberProfile: MemberProfileNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface MemberProfileNavigator {
    data class Args(
        val conversationId: String,
        val memberId: String,
    )

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ShareInviteNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ReactionsNavigator {
    data class Args(
        val conversationId: String,
        val messageId: String,
    )

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface AssistantFilesLinksNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

// Conversation Onboarding

@NavigationTarget
interface SetupProfileNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface InviteAcceptedNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface RequestPushNotificationsNavigator {
    class Args

    fun closed(context: ScreenContext)
}

// Settings Sub-Navigators

@NavigationTarget
interface MyInfoNavigator {
    class Args

    fun navigateTo(quicknameRandomizer: QuicknameRandomizerNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface CustomizeSettingsNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface AssistantSettingsNavigator {
    class Args

    fun present(inviteCodeEntry: InviteCodeEntryNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ConnectionsNavigator {
    class Args

    fun present(connectionGrant: ConnectionGrantNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface BackupRestoreNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface DeleteAllDataNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface QuicknameRandomizerNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface InviteCodeEntryNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ConnectionGrantNavigator {
    data class Args(
        val serviceId: String,
        val conversationId: String,
    )

    fun closed(context: ScreenContext)
}

// Info Sheets

@NavigationTarget
interface ExplodeInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface PinLimitInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface LockedConvoInfoNavigator {
    data class Args(val conversationId: String)

    fun present(lockConfirmation: LockConvoConfirmationNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface FullConvoInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ConversationForkedInfoNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface RevealMediaInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface PhotosInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface AssistantConfirmationNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface AssistantInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ProcessingPowerInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface ExplodedInviteInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface BackwardsSecrecyInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface MaxedOutInfoNavigator {
    class Args

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface LockConvoConfirmationNavigator {
    data class Args(val conversationId: String)

    fun closed(context: ScreenContext)
}

// Paywall & Billing

@NavigationTarget
interface PaywallNavigator {
    data class Args(val source: PaywallSource)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface SubscriptionSettingsNavigator {
    class Args

    fun present(paywall: PaywallNavigator.Args)

    fun closed(context: ScreenContext)
}

@NavigationTarget
interface BillingDebugNavigator {
    class Args

    fun present(paywall: PaywallNavigator.Args)
    fun navigateTo(subscriptionSettings: SubscriptionSettingsNavigator.Args)

    fun closed(context: ScreenContext)
}

enum class PaywallSource {
    SETTINGS,
    LOW_BALANCE_BANNER,
    ONBOARDING,
    MEMBER_CARD,
    DEBUG,
}

File: ConvosMetrics/Sources/ConvosMetrics/NavigationTargets.swift
public enum NewConversationMode {
    case create
    case scanner
    case joinInvite
}

public enum PaywallSource {
    case settings
    case lowBalanceBanner
    case onboarding
    case memberCard
    case debug
}

public struct ConversationsNavigatorArgs {
    public init() {}
}

public struct ConversationNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct AppSettingsNavigatorArgs {
    public init() {}
}

public struct NewConversationNavigatorArgs {
    public let mode: NewConversationMode
    public let inviteCode: String?

    public init(mode: NewConversationMode, inviteCode: String? = nil) {
        self.mode = mode
        self.inviteCode = inviteCode
    }
}

public struct ExplodeConfirmationNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct ConversationInfoNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct ConversationInfoEditNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct MembersListNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct MemberProfileNavigatorArgs {
    public let conversationId: String
    public let memberId: String

    public init(conversationId: String, memberId: String) {
        self.conversationId = conversationId
        self.memberId = memberId
    }
}

public struct ShareInviteNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct ReactionsNavigatorArgs {
    public let conversationId: String
    public let messageId: String

    public init(conversationId: String, messageId: String) {
        self.conversationId = conversationId
        self.messageId = messageId
    }
}

public struct AssistantFilesLinksNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct SetupProfileNavigatorArgs {
    public init() {}
}

public struct InviteAcceptedNavigatorArgs {
    public init() {}
}

public struct RequestPushNotificationsNavigatorArgs {
    public init() {}
}

public struct MyInfoNavigatorArgs {
    public init() {}
}

public struct CustomizeSettingsNavigatorArgs {
    public init() {}
}

public struct AssistantSettingsNavigatorArgs {
    public init() {}
}

public struct ConnectionsNavigatorArgs {
    public init() {}
}

public struct BackupRestoreNavigatorArgs {
    public init() {}
}

public struct DeleteAllDataNavigatorArgs {
    public init() {}
}

public struct QuicknameRandomizerNavigatorArgs {
    public init() {}
}

public struct InviteCodeEntryNavigatorArgs {
    public init() {}
}

public struct ConnectionGrantNavigatorArgs {
    public let serviceId: String
    public let conversationId: String

    public init(serviceId: String, conversationId: String) {
        self.serviceId = serviceId
        self.conversationId = conversationId
    }
}

public struct ExplodeInfoNavigatorArgs {
    public init() {}
}

public struct PinLimitInfoNavigatorArgs {
    public init() {}
}

public struct LockedConvoInfoNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct FullConvoInfoNavigatorArgs {
    public init() {}
}

public struct ConversationForkedInfoNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct RevealMediaInfoNavigatorArgs {
    public init() {}
}

public struct PhotosInfoNavigatorArgs {
    public init() {}
}

public struct AssistantConfirmationNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct AssistantInfoNavigatorArgs {
    public init() {}
}

public struct ProcessingPowerInfoNavigatorArgs {
    public init() {}
}

public struct ExplodedInviteInfoNavigatorArgs {
    public init() {}
}

public struct BackwardsSecrecyInfoNavigatorArgs {
    public init() {}
}

public struct MaxedOutInfoNavigatorArgs {
    public init() {}
}

public struct LockConvoConfirmationNavigatorArgs {
    public let conversationId: String

    public init(conversationId: String) {
        self.conversationId = conversationId
    }
}

public struct PaywallNavigatorArgs {
    public let source: PaywallSource

    public init(source: PaywallSource) {
        self.source = source
    }
}

public struct SubscriptionSettingsNavigatorArgs {
    public init() {}
}

public struct BillingDebugNavigatorArgs {
    public init() {}
}

public protocol ConversationsNavigator: AnyObject {
    func navigateTo(conversation: ConversationNavigatorArgs)
    func present(appSettings: AppSettingsNavigatorArgs)
    func present(newConversation: NewConversationNavigatorArgs)
    func present(explodeConfirmation: ExplodeConfirmationNavigatorArgs)
    func present(connectionGrant: ConnectionGrantNavigatorArgs)
    func present(explodeInfo: ExplodeInfoNavigatorArgs)
    func present(pinLimitInfo: PinLimitInfoNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol ConversationNavigator: AnyObject {
    func present(paywall: PaywallNavigatorArgs)
    func present(conversationInfo: ConversationInfoNavigatorArgs)
    func present(myInfo: MyInfoNavigatorArgs)
    func present(memberProfile: MemberProfileNavigatorArgs)
    func present(shareInvite: ShareInviteNavigatorArgs)
    func present(newConversation: NewConversationNavigatorArgs)
    func present(reactions: ReactionsNavigatorArgs)
    func present(explodeInfo: ExplodeInfoNavigatorArgs)
    func present(lockedConvoInfo: LockedConvoInfoNavigatorArgs)
    func present(fullConvoInfo: FullConvoInfoNavigatorArgs)
    func present(conversationForkedInfo: ConversationForkedInfoNavigatorArgs)
    func present(revealMediaInfo: RevealMediaInfoNavigatorArgs)
    func present(photosInfo: PhotosInfoNavigatorArgs)
    func present(assistantConfirmation: AssistantConfirmationNavigatorArgs)
    func present(assistantInfo: AssistantInfoNavigatorArgs)
    func present(processingPowerInfo: ProcessingPowerInfoNavigatorArgs)
    func present(explodedInviteInfo: ExplodedInviteInfoNavigatorArgs)
    func present(setupProfile: SetupProfileNavigatorArgs)
    func present(inviteAccepted: InviteAcceptedNavigatorArgs)
    func present(requestPushNotifications: RequestPushNotificationsNavigatorArgs)
    func present(backwardsSecrecyInfo: BackwardsSecrecyInfoNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol AppSettingsNavigator: AnyObject {
    func navigateTo(myInfo: MyInfoNavigatorArgs)
    func navigateTo(customize: CustomizeSettingsNavigatorArgs)
    func navigateTo(assistants: AssistantSettingsNavigatorArgs)
    func navigateTo(connections: ConnectionsNavigatorArgs)
    func navigateTo(backupRestore: BackupRestoreNavigatorArgs)
    func navigateTo(deleteAllData: DeleteAllDataNavigatorArgs)
    func navigateTo(subscriptionSettings: SubscriptionSettingsNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol NewConversationNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ExplodeConfirmationNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ConversationInfoNavigator: AnyObject {
    func navigateTo(edit: ConversationInfoEditNavigatorArgs)
    func navigateTo(membersList: MembersListNavigatorArgs)
    func navigateTo(filesAndLinks: AssistantFilesLinksNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol ConversationInfoEditNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol MembersListNavigator: AnyObject {
    func navigateTo(memberProfile: MemberProfileNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol MemberProfileNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ShareInviteNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ReactionsNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol AssistantFilesLinksNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol SetupProfileNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol InviteAcceptedNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol RequestPushNotificationsNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol MyInfoNavigator: AnyObject {
    func navigateTo(quicknameRandomizer: QuicknameRandomizerNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol CustomizeSettingsNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol AssistantSettingsNavigator: AnyObject {
    func present(inviteCodeEntry: InviteCodeEntryNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol ConnectionsNavigator: AnyObject {
    func present(connectionGrant: ConnectionGrantNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol BackupRestoreNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol DeleteAllDataNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol QuicknameRandomizerNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol InviteCodeEntryNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ConnectionGrantNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ExplodeInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol PinLimitInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol LockedConvoInfoNavigator: AnyObject {
    func present(lockConfirmation: LockConvoConfirmationNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol FullConvoInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ConversationForkedInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol RevealMediaInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol PhotosInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol AssistantConfirmationNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol AssistantInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ProcessingPowerInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol ExplodedInviteInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol BackwardsSecrecyInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol MaxedOutInfoNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol LockConvoConfirmationNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol PaywallNavigator: AnyObject {
    func closed(context: ScreenContext)
}

public protocol SubscriptionSettingsNavigator: AnyObject {
    func present(paywall: PaywallNavigatorArgs)
    func closed(context: ScreenContext)
}

public protocol BillingDebugNavigator: AnyObject {
    func present(paywall: PaywallNavigatorArgs)
    func navigateTo(subscriptionSettings: SubscriptionSettingsNavigatorArgs)
    func closed(context: ScreenContext)
}
